File: windows/src/OpenShrimp.Tray/Core/CoreSupervisor.cs
using System.Diagnostics;

namespace OpenShrimp.Tray.Core;

internal enum CoreState { Stopped, Installing, Starting, Running, Stopping, Error, NoConfig }

/// <summary>
/// What a stop established about the core it was asked to stop.
///
/// Separate from <see cref="CoreState"/> because that cannot carry it: a stop
/// ends at <see cref="CoreState.Stopped"/> however the core went, so anything
/// that has to know whether the core is really down — an update about to
/// overwrite its binary — must be told rather than read the state.
/// </summary>
internal enum StopOutcome
{
    /// <summary>
    /// The control endpoint is not answering: it either went quiet inside the
    /// grace period or was already gone when the stop was asked for. The only
    /// outcome that says nothing is still unwinding a sandbox guest, because
    /// the endpoint is held by the interpreter and the handle the tray spawned
    /// is PyApp's launcher.
    /// </summary>
    Quiet,

    /// <summary>
    /// The grace period lapsed with the endpoint still up. The core took the
    /// request and is still working through it.
    /// </summary>
    Lapsed,

    /// <summary>The core answered with an error and was left running.</summary>
    Refused,

    /// <summary>
    /// A core speaking a control protocol this build does not know. Nothing was
    /// sent to it and it is still running.
    /// </summary>
    UnknownProtocol,
}

/// <summary>
/// Owns the core process: starts it, watches it, stops it gracefully.
///
/// Two things drive the design.
///
/// First, the core must never be hard-killed while it holds a sandbox guest —
/// TerminateProcess skips shutdown entirely and strands the HCS compute
/// system. Stopping therefore goes through the control channel, with a kill
/// only as a last resort after the core has had time to unwind.
///
/// Second, the core replaces itself on /restart and on auto-update, so its pid
/// changes underneath us. Liveness is judged by the control endpoint, which
/// keeps its name across the re-exec, rather than by the child handle.
/// </summary>
internal sealed class CoreSupervisor : IAsyncDisposable
{
    private static readonly TimeSpan GracefulStopTimeout = TimeSpan.FromSeconds(45);

    /// <summary>How long a dropped pipe may stay dropped before it counts as gone.</summary>
    private static readonly TimeSpan ReexecGrace = TimeSpan.FromSeconds(30);

    /// <summary>
    /// How long the core gets to open its control channel. It opens the channel
    /// before the rest of its boot, so this bounds process start and config
    /// load, not the whole startup.
    /// </summary>
    private static readonly TimeSpan HandshakeTimeout = TimeSpan.FromSeconds(30);

    /// <summary>
    /// How long the runtime bootstrap may run before the tray reports that it is
    /// installing. Short enough to explain a wait, long enough that an
    /// already-installed core never flashes the message.
    /// </summary>
    private static readonly TimeSpan InstallAnnounceDelay = TimeSpan.FromSeconds(3);

    /// <summary>
    /// Said when a core claims a channel this build does not know. It names the
    /// tray, because the tray is the half that can be fixed from here.
    /// </summary>
    private const string UnknownProtocolDetail = "This core is newer than the tray — update OpenShrimp";

    private readonly string? _instanceName;
    private Process? _process;
    private ControlClient? _client;
    private CancellationTokenSource? _watchdog;
    private CancellationTokenSource? _bootstrap;
    private bool _stopRequested;

    /// <summary>
    /// True once a core has claimed a protocol this build does not know. Every
    /// method on the channel is a guess from then on, including shutdown, so
    /// none are sent.
    /// </summary>
    private bool _unknownProtocol;

    /// <summary>The core's version as last reported, for logging drift.</summary>
    private string? _coreVersion;

    public CoreState State { get; private set; } = CoreState.Stopped;
    public string? StatusDetail { get; private set; }
    public CoreStatus? LastStatus { get; private set; }

    public event Action? Changed;

    public CoreSupervisor(string? instanceName) => _instanceName = instanceName;

    private void Set(CoreState state, string? detail = null)
    {
        State = state;
        StatusDetail = detail;
        Changed?.Invoke();
    }

    public async Task StartAsync()
    {
        if (State is CoreState.Running or CoreState.Starting or CoreState.Installing) return;

        if (!File.Exists(CorePaths.ConfigFile))
        {
            Set(CoreState.NoConfig);
            return;
        }

        _stopRequested = false;
        // Describes the core that was attached last time, and there is none
        // attached now. Left standing, it would make StopAsync walk away from a
        // core this start is about to spawn.
        _unknownProtocol = false;
        Set(CoreState.Starting);

        // Adopt a core that is already running — started from a terminal, or
        // left behind by a tray that was killed. Starting a second one would
        // collide on the control endpoint anyway.
        var adopted = new ControlClient(_instanceName);
        if (await adopted.TryConnectAsync(TimeSpan.FromSeconds(1)).ConfigureAwait(false))
        {
            AttachClient(adopted);
            await RefreshStatusAsync().ConfigureAwait(false);
            StartWatchdog();
            return;
        }
        await adopted.DisposeAsync().ConfigureAwait(false);

        var exe = CorePaths.CoreExecutable;
        if (!File.Exists(exe))
        {
            Set(CoreState.Error, $"Core executable not found at {exe}");
            return;
        }

        // A core we spawned before and never reached is still unreachable — the
        // adopt probe above just failed against the endpoint it would hold.
        // Retiring it here rather than when its handshake timed out is what
        // lets a merely slow core keep running and be adopted instead.
        RetireUnreachableProcess();

        // Unpack the runtime before anything starts timing the boot. Killing a
        // core midway through installing itself leaves it permanently broken,
        // so this step is deliberately unbounded — a stop stops us waiting on
        // it, and lets it run to completion unsupervised.
        _bootstrap = new CancellationTokenSource();
        var bootstrap = OpenShrimpCli.EnsureRuntimeAsync(_bootstrap.Token);
        if (await Task.WhenAny(bootstrap, Task.Delay(InstallAnnounceDelay)).ConfigureAwait(false) != bootstrap)
            Set(CoreState.Installing);
        var bootstrapError = await bootstrap.ConfigureAwait(false);
        if (_stopRequested)
        {
            Set(CoreState.Stopped);
            return;
        }
        if (bootstrapError is not null)
        {
            Set(CoreState.Error, bootstrapError);
            return;
        }
        Set(CoreState.Starting);

        try
        {
            var startInfo = new ProcessStartInfo
            {
                FileName = exe,
                Arguments = $"--config \"{CorePaths.ConfigFile}\"",
                UseShellExecute = false,
                CreateNoWindow = true,
                // The core owns its own rotating log file, so its console
                // streams are left alone rather than pumped into a second one.
            };
            // Something above the core owns keeping it alive — this supervisor
            // while the tray runs, and its logon task across sign-ins — so the
            // core must not ask the operator to register one of its own.
            startInfo.Environment["OPENSHRIMP_SUPERVISED"] = "1";
            // This tray installs both halves of a release: one MSI carries the
            // tray and the core, and the tray hands it to msiexec off a signed
            // feed. A core polling GitHub as well would offer a second Update
            // button for the same release and write over the binary the MSI
            // replaces.
            startInfo.Environment["OPENSHRIMP_UPDATES_MANAGED"] = "1";
            _process = Process.Start(startInfo);
        }
        catch (Exception ex)
        {
            Set(CoreState.Error, ex.Message);
            return;
        }

        var client = new ControlClient(_instanceName);
        if (!await client.TryConnectAsync(HandshakeTimeout).ConfigureAwait(false))
        {
            await client.DisposeAsync().ConfigureAwait(false);
            // Leave it running rather than killing it: a core that is only slow
            // gets adopted on the next Start, and a kill here would cut it off
            // mid-write. The handle is kept so that Start can retire it first
            // instead of orphaning a core that still holds the control endpoint
            // and keeps polling Telegram unsupervised.
            Set(CoreState.Error, "Core did not open its control channel");
            return;
        }

        AttachClient(client);
        await RefreshStatusAsync().ConfigureAwait(false);
        StartWatchdog();
    }

    private void AttachClient(ControlClient client)
    {
        _client = client;
        client.EventReceived += OnCoreEvent;
        client.Disconnected += OnDisconnected;
    }

    private void OnCoreEvent(string name, System.Text.Json.JsonElement? data)
    {
        switch (name)
        {
            case "state":
                _ = RefreshStatusAsync();
                break;
            case "stopping":
                // A stop we did not ask for is a restart, not a crash.
                Set(CoreState.Stopping);
                break;
        }
    }

    private void OnDisconnected()
    {
        if (_stopRequested)
        {
            Set(CoreState.Stopped);
            return;
        }

        // The core may simply be re-execing. Give the endpoint a chance to
        // come back before calling it dead.
        _ = Task.Run(async () =>
        {
            var client = _client;
            if (client is not null && await client.ReconnectAsync(ReexecGrace).ConfigureAwait(false))
            {
                // The core re-execed, so the handle we spawned refers to a dead
                // pid. Keeping it would make the graceful-stop wait exit on its
                // first tick against an already-exited process, reporting the
                // live core as stopped and leaving the kill fallback dead too.
                DropStaleProcessHandle();
                if (_stopRequested) return;
                await RefreshStatusAsync().ConfigureAwait(false);
                return;
            }
            if (!_stopRequested) Set(CoreState.Error, "Core stopped unexpectedly");
        });
    }

    /// <summary>
    /// Kill a core that was spawned but never answered on the control channel,
    /// before spawning its replacement. Only ever called once the adopt probe
    /// has failed, which is what establishes that it is unreachable and not
    /// merely slow.
    /// </summary>
    private void RetireUnreachableProcess()
    {
        var process = _process;
        _process = null;
        if (process is null) return;

        try
        {
            if (!process.HasExited) process.Kill(entireProcessTree: true);
        }
        catch (Exception) { /* already gone */ }
        process.Dispose();
    }

    /// <summary>
    /// Release the spawned handle once it no longer refers to the live core.
    /// After this the control endpoint is the only liveness signal, which is
    /// the same footing the adopt path runs on.
    /// </summary>
    private void DropStaleProcessHandle()
    {
        var process = _process;
        _process = null;
        process?.Dispose();
    }

    public async Task RefreshStatusAsync()
    {
        var client = _client;
        if (client is null || !client.IsConnected) return;

        var status = await client.GetStatusAsync().ConfigureAwait(false);
        if (status is null) return;

        LastStatus = status;

        // Checked before the state below is mapped: a core speaking a channel
        // this build does not know may mean something else by every field in the
        // reply, and by every method the tray could call back on. Saying so is
        // the last thing the tray can still do correctly.
        if (status.Protocol > ControlProtocol.Expected)
        {
            if (!_unknownProtocol)
                TrayLog.Write($"Core speaks control protocol {status.Protocol}; " +
                              $"this build knows {ControlProtocol.Expected}");
            _unknownProtocol = true;
            _coreVersion = null;
            Set(CoreState.Error, UnknownProtocolDetail);
            return;
        }
        _unknownProtocol = false;

        // Recorded, not shown. The core self-replacing past the version the MSI
        // installed is how Windows updates at all, so a tray that called that a
        // fault would be flagging the normal case forever — but which version
        // is actually running is still the first thing anyone diagnosing this
        // needs, and it is nowhere else.
        if (status.Version != _coreVersion)
        {
            _coreVersion = status.Version;
            TrayLog.Write($"Core is version {_coreVersion ?? "unreported"}; " +
                          $"this build is {TrayVersion.Current}");
        }

        Set(status.State switch
        {
            "running" => CoreState.Running,
            "starting" => CoreState.Starting,
            "stopping" => CoreState.Stopping,
            "error" => CoreState.Error,
            _ => CoreState.Starting,
        }, status.Error ?? status.BotUsername);
    }

    private void StartWatchdog()
    {
        _watchdog?.Cancel();
        _watchdog = new CancellationTokenSource();
        var ct = _watchdog.Token;
        _ = Task.Run(async () =>
        {
            while (!ct.IsCancellationRequested)
            {
                await Task.Delay(TimeSpan.FromSeconds(10), ct).ConfigureAwait(false);
                await RefreshStatusAsync().ConfigureAwait(false);
            }
        }, ct);
    }

    /// <summary>
    /// Stop the core, and say what that established. The answer is the caller's
    /// only evidence: <see cref="State"/> ends at <see cref="CoreState.Stopped"/>
    /// on every path below, including the ones that leave a core running.
    /// </summary>
    public async Task<StopOutcome> StopAsync()
    {
        _stopRequested = true;
        _watchdog?.Cancel();
        // Stop waiting on a runtime install; never interrupt one. The command
        // outlives the wait and finishes on its own, which is what keeps its
        // installation directory from being left half-written.
        _bootstrap?.Cancel();

        // A core speaking a channel this build does not know is left running.
        // Shutdown may not mean that any more, and the only fallback here is
        // TerminateProcess — which is how an HCS guest gets stranded. Walking
        // away strands nothing: the core is still supervising itself, and the
        // next front end to meet it will be one that can speak to it.
        if (_unknownProtocol)
        {
            TrayLog.Write("Leaving a core that speaks an unknown control protocol running");
            await TeardownAsync().ConfigureAwait(false);
            Set(CoreState.Stopped);
            return StopOutcome.UnknownProtocol;
        }

        Set(CoreState.Stopping);

        var client = _client;
        var connected = client is not null && client.IsConnected;

        // How the core went is the difference between a sandbox guest that
        // unwound and one that was stranded, and this is the only account of it
        // anyone gets. It matters most at the end of a session, where the core
        // can already be gone before the tray is asked to stop it.
        TrayLog.Write($"Stopping the core: control channel {(connected ? "up" : "gone")}, " +
                      $"process {(_process is null ? "not ours" : _process.HasExited ? "already exited" : "running")}");

        // Quiet until something says otherwise: a stop asked of a core that was
        // already gone, or of one that was never attached, has established
        // exactly what a successful stop establishes. The endpoint not
        // answering is the whole of the claim.
        var outcome = StopOutcome.Quiet;

        if (connected)
        {
            // The reply is read rather than merely awaited, because an error
            // frame is still a frame: a core that no longer implements the
            // method answers unknown_method. Silence is a different matter and
            // is not read as a refusal — a core wedged in its teardown answers
            // nothing, and it is the one that most needs the grace period.
            var reply = await client!.ShutdownAsync().ConfigureAwait(false);
            if (reply?.Error is not null)
            {
                // It answered and declined. There is no drain to wait for and
                // no second way to ask, and the only rung left here is
                // TerminateProcess — which on a core holding an HCS compute
                // system strands the guest, so it is not taken. Left running,
                // like Stop.cs leaves one it cannot drain.
                TrayLog.Write($"Core refused the stop request ({reply.Error.Code}: " +
                              $"{reply.Error.Message}); leaving it running");
                outcome = StopOutcome.Refused;
            }
            else
            {
                // Shutdown has no internal timeout on the Python side — a wedged
                // sandbox teardown can hang it — so bound the wait here.
                //
                // The control channel is what says the core is down, and the
                // spawned handle says nothing: the image the tray launches is
                // PyApp's, and it is a different process from the interpreter that
                // actually holds the sandbox. That distinction is load-bearing at
                // the end of a session, where the interpreter defers itself to the
                // back of the shutdown order and the launcher does not — so the
                // launcher exits first, every time, while the guest is still
                // unwinding. Breaking on it would report a stopped core and let go
                // of the session the drain is holding open.
                var deadline = DateTime.UtcNow + GracefulStopTimeout;
                while (DateTime.UtcNow < deadline && client.IsConnected)
                    await Task.Delay(250).ConfigureAwait(false);

                if (client.IsConnected)
                {
                    TrayLog.Write("Core did not answer the stop within its grace period");
                    outcome = StopOutcome.Lapsed;
                }
            }
        }

        // Said out loud because the kill below cannot reach it. Killing the
        // launcher's tree is no use once the launcher has gone, which at the
        // end of a session it has, so a core still holding a guest here is
        // one nothing is going to unwind.
        if (_process is null or { HasExited: true } && _client is { IsConnected: true })
            TrayLog.Write("Core is still up with no handle to stop it by");

        if (_process is { HasExited: false } && outcome != StopOutcome.Refused)
        {
            // The outcome this whole class is arranged to avoid: a core killed
            // while it holds an HCS compute system strands the guest.
            TrayLog.Write("Core did not stop in time; killing it");
            try { _process.Kill(entireProcessTree: true); }
            catch (Exception) { /* already gone */ }
        }

        await TeardownAsync().ConfigureAwait(false);
        Set(CoreState.Stopped);
        return outcome;
    }

    private async Task TeardownAsync()
    {
        var client = _client;
        _client = null;
        if (client is not null)
        {
            client.EventReceived -= OnCoreEvent;
            client.Disconnected -= OnDisconnected;
            await client.DisposeAsync().ConfigureAwait(false);
        }
        _process?.Dispose();
        _process = null;
        // Everything read off a core describes the one that was attached, so it
        // goes when the attachment does.
        LastStatus = null;
        _coreVersion = null;
        _unknownProtocol = false;
    }

    public async ValueTask DisposeAsync()
    {
        _watchdog?.Cancel();
        await TeardownAsync().ConfigureAwait(false);
    }
}

File: macos/Sources/OpenShrimpCore/CoreSupervisor.swift
import Foundation

enum CoreState: String, Sendable {
    case stopped, installing, starting, running, stopping, error, noConfig
}

/// Owns the core process: starts it, watches it, stops it gracefully.
///
/// Two things drive the design.
///
/// First, the core must never be hard-killed while it holds a sandbox guest.
/// SIGKILL skips shutdown entirely and strands a Lima VM and a cloudflared
/// tunnel, with a `sessions.db` write possibly still in flight.  Stopping
/// therefore goes through the control channel, and signals are an escalation
/// after the core has had time to unwind.
///
/// Second, the core replaces itself on `/restart` and on auto-update, so its
/// pid changes underneath us.  Liveness is judged by the control endpoint,
/// which keeps its path across the re-exec, rather than by the child handle.
actor CoreSupervisor {
    /// `shutdown` has no internal timeout on the Python side — a wedged sandbox
    /// teardown can hang it — so the wait is bounded here.
    static let gracefulStopTimeout: TimeInterval = 45

    /// How long a dropped socket may stay dropped before it counts as gone.
    static let reexecGrace: TimeInterval = 30

    /// How long the core gets to open its control channel.  It opens the
    /// channel before the rest of its boot, so this bounds process start and
    /// config load, not the whole startup.
    static let handshakeTimeout: TimeInterval = 30

    /// How long the runtime bootstrap may run before the app reports that it is
    /// installing.  Short enough to explain a wait, long enough that an
    /// already-installed core never flashes the message.
    static let installAnnounceDelay: TimeInterval = 3

    static let watchdogPeriod: TimeInterval = 10

    private static let gracefulStopPoll: TimeInterval = 0.25

    /// How long SIGTERM gets before SIGKILL.
    ///
    /// Shorter than the graceful budget rather than a second copy of it: the
    /// core installs a SIGTERM handler that re-enters the same shutdown the RPC
    /// already asked for, so reaching this rung means either that shutdown
    /// never started or that it is already wedged.  This rung has no Windows
    /// equivalent, where the only kill available is unconditional.
    private static let sigtermGrace: TimeInterval = 10

    private let instanceName: String?
    private var process: Process?
    private var client: ControlClient?
    private var watchdog: Task<Void, Never>?
    private var bootstrap: Task<String?, Never>?
    private var stopRequested = false

    private(set) var state: CoreState = .stopped
    private(set) var statusDetail: String?
    private(set) var lastStatus: CoreStatus?

    /// Everything a caller outside the actor renders, read in a single hop.
    /// Three separate reads can interleave with a watchdog poll and produce a
    /// line that pairs a state from one moment with a detail from another.
    struct Snapshot: Sendable {
        let state: CoreState
        let detail: String?
        let botUsername: String?
    }

    func snapshot() -> Snapshot {
        Snapshot(state: state, detail: statusDetail, botUsername: lastStatus?.botUsername)
    }

    private var onChange: (@Sendable (CoreState, String?) -> Void)?

    init(instanceName: String?) {
        self.instanceName = instanceName
    }

    func setOnChange(_ handler: (@Sendable (CoreState, String?) -> Void)?) {
        onChange = handler
    }

    /// Announces only what changed.  The watchdog re-reads status on a timer,
    /// so a handler told about every read would redraw a menu every tick and
    /// bury the transitions that matter among identical lines.
    private func set(_ state: CoreState, _ detail: String? = nil) {
        guard state != self.state || detail != statusDetail else { return }
        self.state = state
        statusDetail = detail
        onChange?(state, detail)
    }

    // -- Start ---------------------------------------------------------------

    func start() async {
        if state == .running || state == .starting || state == .installing { return }

        guard FileManager.default.fileExists(atPath: CorePaths.configFile.path) else {
            set(.noConfig)
            return
        }

        stopRequested = false
        set(.starting)

        // Adopt a core that is already running — started from a terminal, or
        // left behind by a front end that was killed.  Starting a second one
        // would collide on the control endpoint anyway.
        let adopted = ControlClient(instanceName: instanceName)
        guard !adopted.endpointExceedsSunPath else {
            // Nothing can ever listen at this address, so reporting it as a
            // core that is not running would send the user looking in the
            // wrong place forever.
            set(.error, "Control socket path is too long — shorten instance_name")
            return
        }
        if await adopted.connect() {
            await attach(adopted)
            await refreshStatus()
            startWatchdog()
            return
        }
        await adopted.dispose()

        if let reason = CorePaths.seedCoreIfNeeded() {
            set(.error, reason)
            return
        }

        let executable = CorePaths.coreExecutable
        guard FileManager.default.isExecutableFile(atPath: executable.path) else {
            set(.error, "Core executable not found at \(executable.path)")
            return
        }

        // A core we spawned before and never reached is still unreachable — the
        // adopt probe above just failed against the endpoint it would hold.
        // Retiring it here rather than when its handshake timed out is what
        // lets a merely slow core keep running and be adopted instead.
        await retireUnreachableProcess()

        // Unpack the runtime before anything starts timing the boot.  Killing a
        // core midway through installing itself leaves it permanently broken,
        // so this step is deliberately unbounded — a stop stops us waiting on
        // it, and lets it run to completion unsupervised.
        let bootstrap = Task { await OpenShrimpCLI.ensureRuntime() }
        self.bootstrap = bootstrap
        let announce = Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(Self.installAnnounceDelay * 1_000_000_000))
            guard !Task.isCancelled else { return }
            await self?.announceInstalling()
        }
        let bootstrapError = await bootstrap.value
        announce.cancel()
        self.bootstrap = nil

        if stopRequested {
            set(.stopped)
            return
        }
        if let bootstrapError {
            set(.error, bootstrapError)
            return
        }
        set(.starting)

        let process = Process()
        process.executableURL = executable
        process.arguments = ["--config", CorePaths.configFile.path]
        // The core owns its own rotating log file, so its output streams are
        // left alone rather than pumped into a second one.
        do {
            try process.run()
        } catch {
            set(.error, error.localizedDescription)
            return
        }
        self.process = process

        let client = ControlClient(instanceName: instanceName)
        guard await client.awaitConnection(within: Self.handshakeTimeout) else {
            await client.dispose()
            // Leave it running rather than killing it: a core that is only slow
            // gets adopted on the next start, and a kill here would cut it off
            // mid-write.  The handle is kept so that the next start can retire
            // it first, instead of orphaning a core that still holds the
            // control endpoint and keeps polling Telegram unsupervised.
            set(.error, "Core did not open its control channel")
            return
        }

        await attach(client)
        await refreshStatus()
        startWatchdog()
    }

    private func announceInstalling() {
        guard state == .starting else { return }
        set(.installing)
    }

    private func attach(_ client: ControlClient) async {
        self.client = client
        await client.setHandlers(
            onEvent: { [weak self] event in
                Task { await self?.handle(event: event) }
            },
            onDisconnect: { [weak self] in
                Task { await self?.handleDisconnect() }
            }
        )
    }

    private func handle(event: String) async {
        switch event {
        case "state":
            await refreshStatus()
        case "stopping":
            // A stop we did not ask for is a restart, not a crash.
            set(.stopping)
        default:
            break
        }
    }

    private func handleDisconnect() async {
        if stopRequested {
            set(.stopped)
            return
        }

        // The core may simply be re-execing.  Give the endpoint a chance to
        // come back before calling it dead.
        guard let client else {
            set(.error, "Core stopped unexpectedly")
            return
        }

        if await client.awaitConnection(within: Self.reexecGrace) {
            reconcileProcessHandle()
            if stopRequested { return }
            await refreshStatus()
            return
        }
        if !stopRequested { set(.error, "Core stopped unexpectedly") }
    }

    /// Retire a core that was spawned but never answered on the control
    /// channel, before spawning its replacement.  Only ever called once the
    /// adopt probe has failed, which is what establishes that it is
    /// unreachable and not merely slow.
    ///
    /// Waited on rather than left to finish in the background: an old core that
    /// opened its endpoint late would make the replacement refuse to start,
    /// and the user would see a spawn fail for a reason nothing on screen
    /// explains.
    private func retireUnreachableProcess() async {
        guard let process else { return }
        self.process = nil
        guard process.isRunning else { return }

        // SIGTERM first even here, because the core is what shuts down the
        // Lima VM and the tunnel it spawned; SIGKILL leaves both running with
        // nothing left to own them.
        //
        // Signalled by pid rather than through `Process.terminate()`, which
        // raises when the process has already gone; `kill` merely reports
        // `ESRCH`.
        let pid = process.processIdentifier
        guard pid > 0 else { return }
        kill(pid, SIGTERM)
        await waitForExit(of: pid, deadline: Date().addingTimeInterval(Self.sigtermGrace))
        if Self.isAlive(pid) { kill(pid, SIGKILL) }
    }

    /// Decide, after a reconnect, whether the spawned handle still refers to
    /// the core now answering.
    ///
    /// A re-exec here is `os.execv`, which replaces the process image and keeps
    /// the pid, so the handle usually still owns the live core — and keeping it
    /// is what preserves the signal fallback for a later stop.  Windows has no
    /// equivalent: there the restart spawns a child and the old pid dies, so
    /// the handle is always stale.
    ///
    /// The case that must still drop it is a core that really did exit and had
    /// its endpoint taken over by one we did not spawn.  A dead handle would
    /// make the graceful-stop wait exit on its first tick, reporting a live
    /// core as stopped and leaving the kill fallback aimed at nothing.
    private func reconcileProcessHandle() {
        guard let process, !process.isRunning else { return }
        self.process = nil
    }

    // -- Status --------------------------------------------------------------

    func refreshStatus() async {
        guard let client, await client.isConnected else { return }
        guard let status = await client.status() else { return }

        lastStatus = status
        let mapped: CoreState
        switch status.state {
        case "running": mapped = .running
        case "starting": mapped = .starting
        case "stopping": mapped = .stopping
        case "error": mapped = .error
        default: mapped = .starting
        }
        set(mapped, status.error ?? status.botUsername)
    }

    private func startWatchdog() {
        watchdog?.cancel()
        watchdog = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: UInt64(Self.watchdogPeriod * 1_000_000_000))
                guard !Task.isCancelled else { return }
                await self?.refreshStatus()
            }
        }
    }

    // -- Stop ----------------------------------------------------------------

    func stop() async {
        stopRequested = true
        watchdog?.cancel()
        watchdog = nil
        // Stop waiting on a runtime install; never interrupt one.  The command
        // outlives the wait and finishes on its own, which is what keeps its
        // installation directory from being left half-written.
        bootstrap?.cancel()
        bootstrap = nil
        set(.stopping)

        if let client, await client.isConnected {
            _ = await client.shutdown()
            await waitForStop(deadline: Date().addingTimeInterval(Self.gracefulStopTimeout))
        }

        if let target = await escalationTarget() {
            // The core installs a SIGTERM handler, so this asks again through
            // the one other channel it listens on before giving up on a clean
            // unwind.
            kill(target, SIGTERM)
            await waitForExit(of: target, deadline: Date().addingTimeInterval(Self.sigtermGrace))
            if Self.isAlive(target) { kill(target, SIGKILL) }
        }

        await teardown()
        set(.stopped)
    }

    /// The pid an escalation may signal, or nil when there is none we can
    /// justify signalling.
    ///
    /// A core we spawned is obvious.  An adopted one is not, and is the common
    /// case: the app is routinely started after a core launched from a
    /// terminal.  Leaving that one unstoppable would strand exactly the Lima VM
    /// and tunnel this path exists to shut down — so its own reported pid is
    /// used, but only while its socket is still open.  The open socket is what
    /// proves the pid still belongs to the core; without it the number may
    /// since have been reused by something else entirely.
    private func escalationTarget() async -> pid_t? {
        if let process, process.isRunning { return process.processIdentifier }
        guard let client, await client.isConnected else { return nil }
        guard let pid = lastStatus?.pid, pid > 0 else { return nil }
        return pid_t(pid)
    }

    private static func isAlive(_ pid: pid_t) -> Bool {
        kill(pid, 0) == 0 || errno == EPERM
    }

    /// Wait until the core is gone, judged by the spawned handle when we have
    /// one and by the control endpoint when we do not — an adopted core, or one
    /// that re-execed out from under its handle, has no handle to watch.
    private func waitForStop(deadline: Date) async {
        while Date() < deadline {
            if let process {
                if !process.isRunning { return }
            } else if let client {
                if await client.isConnected == false { return }
            } else {
                return
            }
            try? await Task.sleep(nanoseconds: UInt64(Self.gracefulStopPoll * 1_000_000_000))
        }
    }

    private func waitForExit(of pid: pid_t, deadline: Date) async {
        while Date() < deadline, Self.isAlive(pid) {
            try? await Task.sleep(nanoseconds: UInt64(Self.gracefulStopPoll * 1_000_000_000))
        }
    }

    private func teardown() async {
        if let client {
            await client.setHandlers(onEvent: nil, onDisconnect: nil)
            await client.dispose()
        }
        client = nil
        process = nil
    }

    func dispose() async {
        watchdog?.cancel()
        watchdog = nil
        await teardown()
    }
}
